api_types.ts

```diff
@@ -546,6 +546,20 @@ export interface CommonPackFormulaDef<T extends ParamDefs> {
    * @hidden
    */
   readonly allowedAuthenticationNames?: string[];
+
+  /**
+   * The JavaScript function that implements parameter validation
+   *
+   * This function takes in parameters and a context containing permissions mode information,
+   * and validates the parameters.
+   *
+   * @returns if the parameters are valid
+   * throws {@link ParameterValidationError} if the parameters are invalid.
+   */
+  validateParameters?(
```

Comment:
You'll need to update something or other in metadata.ts. We generate a pure-JSON metadata file from the full pack def, which is what the client and various parts of the app use to know the contents of the pack without looking needing the executable code. So we need bits in the metadata JSON to know whether a given formula has a validator or not. If we wind up making this a metadata formula, you can reuse `compileMetadataFormulaMetadata()`,  otherwise you'll maybe want to do something similar to executeUpdate where we convert that to a `supportUpdates` boolean.